src/HasteModuleLoader/HasteModuleLoader.js

```diff
@@ -159,8 +159,8 @@ class Loader {
   requireMock(currPath, moduleName) {
     const moduleID = this._getNormalizedModuleID(currPath, moduleName);
 
-    if (moduleID in this._explicitlySetMocks) {
-      return this._explicitlySetMocks[moduleID];
+    if (moduleID in this._explicitlySetMockFactories) {
+      return this._explicitlySetMockFactories[moduleID]();
```

Comment:
I think we need to make sure we only run this function once and that we cache the result. Otherwise we'll re-evaluate the mock every time it is being required. I can do that in a follow-up though :)